### PR TITLE
feat: add userop simulation <WIP>

### PR DESCRIPTION
**Pull request by avasisht23**
### Detailed summary
- Added `isValidRequest` function in `packages/core/src/index.ts`.
- Exported several modules and types in `packages/alchemy/src/client/index.ts` and `packages/alchemy/src/index.ts`.
- Added `DebugCallType` and `DecodingAuthority` enums in `packages/alchemy/src/client/enums.ts`.
- Added `simulateUserOperationExecution` method in `packages/alchemy/src/provider.ts`.
- Added `createAlchemyEnhanced4337Client` function in `packages/alchemy/src/client/factory.ts`.
- Added several types and interfaces in `packages/alchemy/src/client/types.ts` and `packages/alchemy/src/client/interfaces.ts`.

> The following files were skipped due to too many changes: `packages/alchemy/src/client/interfaces.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by avasisht23** [comment]
Current dependencies on/for this PR:
* main
  * **PR #175** <a href="https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/175" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/alchemyplatform/aa-sdk/175?utm_source=stack-comment).

File: packages/alchemy/src/client/factory.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+       * If the factory accesses global state, it must be staked - see reputation, throttling and banning section for details."
+       */
+      if (size(initCode) !== 42) {
+        throw new Error(
+          `Invalid init code provided. Expected 0x or 42 bytes, got ${initCode}`
+        );
+      }
+
+      const factoryAddress = slice(initCode, 2, 22);
+      const initCalldata = slice(initCode, 22);
```
Comment (moldy530) [inline packages/alchemy/src/client/factory.ts:67]: we have a method on base account for this already. we should break out the contents of the base account impl into a util function and use that here

File: packages/alchemy/src/provider.ts [hunk 2]
```diff
@@ -127,6 +134,38 @@ export class AlchemyProvider extends SmartAccountProvider<HttpTransport> {
...
   };
 
+  simulateUserOperationExecution = async (
+    uoCallData: UserOperationCallData
+  ): Promise<SimulateExecutionResponse | SimulateExecutionResponse[]> => {
+    const uoStruct = await this.buildUserOperation(uoCallData);
+
+    if (!this.account) {
+      throw new Error("account not connected");
+    }
```
Comment (moldy530) [inline packages/alchemy/src/provider.ts:144]: shouldn't this happen before building?

File: packages/core/src/utils/validation.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+): request is UserOperationRequest {
+  // These are the only ones marked as optional in the interface above
+  return (
+    !!request.callGasLimit &&
+    !!request.maxFeePerGas &&
+    request.maxPriorityFeePerGas != null &&
+    !!request.preVerificationGas &&
+    !!request.verificationGasLimit
+  );
+}
```
Comment (moldy530) [inline packages/core/src/utils/validation.ts:20]: could use `zod` here too

File: packages/alchemy/src/client/factory.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      }
+
+      /**
+       * @see {https://eips.ethereum.org/EIPS/eip-4337#client-behavior-upon-receiving-a-useroperation}
+       *
+       * "If initCode is not empty, parse its first 20 bytes as a factory address.
+       * Record whether the factory is staked, in case the later simulation indicates that it needs to be.
+       * If the factory accesses global state, it must be staked - see reputation, throttling and banning section for details."
+       */
+      if (size(initCode) !== 42) {
```
Comment (moldy530) [inline packages/alchemy/src/client/factory.ts:60]: initCode can be `> 42` in length, it should be minimum 42 in length

File: packages/alchemy/src/client/factory.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        to: sender,
+        data: callData,
+      };
+
+      if (initCode === "0x") {
+        return clientAdapter.request({
+          method: "alchemy_simulateExecution",
+          params: [innerTransactionObject],
+        });
+      }
```
Comment (denniswon) [inline packages/alchemy/src/client/factory.ts:51]: https://docs.alchemy.com/reference/alchemy-simulateexecution

Does this api support BASE chain or not? Either this linked doc is outdated, or if not, we should add a check if the chain is Base chain until the api gets updated with Base support.

File: packages/alchemy/src/client/factory.ts (hunk 1 above)
Comment (denniswon) [reply packages/alchemy/src/client/factory.ts:67]: https://github.com/alchemyplatform/aa-sdk/blob/e7a39f50a71aa21aa7ae73e3d704c4a6709aeb18/packages/core/src/account/base.ts#L338

should create a common util to be used for both.

File: packages/alchemy/src/client/factory.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+ * The following client factory method to a viem public client.
+ *
+ * It extends the client with the following methods:
+ * - `simulateUserOperationExecution`
+ *
+ * @param chain - the chain to connect to
+ * @param rpcUrl - the rpc url to
+ * @param fetchOptions - (optional) the fetch options to use when making requests
+ */
+export const createAlchemyEnhanced4337Client = ({
```
Comment (denniswon) [inline packages/alchemy/src/client/factory.ts:96]: Can we name shorted as `Alchemy4337Client`?

File: packages/alchemy/src/client/factory.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  type HttpTransportConfig,
+  type Transport,
+} from "viem";
+import type { DebugTransaction, SimulateExecutionResponse } from "./interfaces";
+import type {
+  AlchemyEnhanced4337Client,
+  AlchemyEnhancedApiSchema,
+} from "./types";
+
+export const alchemyEnhancedApiActions = (client: Client) => {
```
Comment (denniswon) [inline packages/alchemy/src/client/factory.ts:24]: shorter name as `alchemy4337Actions`?

File: packages/alchemy/src/client/factory.ts (hunk 1 above)
Comment (avasisht23) [reply packages/alchemy/src/client/factory.ts:67]: ah right i forgot about this. will refactor

File: packages/alchemy/src/client/factory.ts (hunk 5 above)
Comment (avasisht23) [reply packages/alchemy/src/client/factory.ts:51]: good point, lemme check with Malak

**Comment by avasisht23** [comment]
closing in favor of #250

**Pull request closed without merging**